File: internal/color/color_test.go
package color

import "testing"

var terminalBackground = RGB{12, 12, 12}

func TestHeadroomLevel(t *testing.T) {
	tests := []struct {
		left float64
		want Level
	}{{30, Green}, {29, Yellow}, {15, Yellow}, {14, Red}, {-5, Red}}
	for _, tt := range tests {
		if got := HeadroomLevel(tt.left, 30, 15); got != tt.want {
			t.Errorf("HeadroomLevel(%v) = %v, want %v", tt.left, got, tt.want)
		}
	}
}

func TestOklchToRGB(t *testing.T) {
	if got, _ := OklchToRGB(1, 0, 0); got != (RGB{255, 255, 255}) {
		t.Errorf("white = %v", got)
	}
	if got, _ := OklchToRGB(0, 0, 0); got != (RGB{}) {
		t.Errorf("black = %v", got)
	}
	if _, ok := OklchToRGB(0.8, 0.4, 265); ok {
		t.Error("an out-of-gamut color was accepted")
	}
	// Regression: the PowerShell version rounded to 0 or 255 before gamma.
	got, _ := OklchToRGB(0.8, 0.1, 255)
	for _, c := range got {
		if c == 0 || c == 255 {
			t.Fatalf("channel clipped: %v", got)
		}
	}
}

// The JavaScript version's colors, so existing users keep their project colors.
var jsSlotColors = [Slots]RGB{
	{252, 160, 167}, {254, 164, 124}, {242, 175, 72}, {203, 194, 74},
	{150, 210, 115}, {81, 218, 167}, {42, 215, 215}, {73, 207, 252},
	{146, 193, 253}, {183, 181, 252}, {224, 161, 251}, {253, 151, 210},
}

func TestSlotColors(t *testing.T) {
	seen := map[RGB]bool{}
	minLum, maxLum := 1.0, 0.0
	for slot := range Slots {
		c := SlotColor(slot)
		if c != jsSlotColors[slot] {
			t.Errorf("slot %d = %v, want %v", slot, c, jsSlotColors[slot])
		}
		if r := ContrastRatio(c, terminalBackground); r < 9 {
			t.Errorf("slot %d contrast %.1f:1 on a dark terminal, want at least 9:1", slot, r)
		}
		seen[c] = true
		lum := RelativeLuminance(c)
		minLum, maxLum = min(minLum, lum), max(maxLum, lum)
	}
	if len(seen) != Slots {
		t.Errorf("only %d distinct colors", len(seen))
	}
	if maxLum/minLum >= 1.4 {
		t.Errorf("brightness varies too much across slots: %.2f", maxLum/minLum)
	}
}

func TestProjectSlot(t *testing.T) {
	if ProjectSlot("TechCentral", nil) != ProjectSlot("techcentral", nil) {
		t.Error("slot depends on case")
	}
	// Same hash as the JavaScript version.
	for name, want := range map[string]int{"techcentral": 7, "orc-pack": 5, "Roga": 9, "api": 6, "sandbox": 3} {
		if got := ProjectSlot(name, nil); got != want {
			t.Errorf("ProjectSlot(%q) = %d, want %d", name, got, want)
		}
	}
	overrides := map[string]int{"website": 4, "api": 13, "neg": -1}
	for name, want := range map[string]int{"Website": 4, "api": 1, "neg": 11} {
		if got := ProjectSlot(name, overrides); got != want {
			t.Errorf("ProjectSlot(%q, overrides) = %d, want %d", name, got, want)
		}
	}
}

File: test/color.test.js
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';
import {
  contrastRatio,
  headroomLevel,
  oklchToRgb,
  PROJECT_SLOTS,
  projectSlot,
  relativeLuminance,
  slotColor,
} from '../src/color.js';

const TERMINAL_BG = [12, 12, 12];

describe('headroomLevel', () => {
  it('uses the configured thresholds', () => {
    const t = { green: 30, yellow: 15 };
    assert.equal(headroomLevel(30, t), 'green');
    assert.equal(headroomLevel(29, t), 'yellow');
    assert.equal(headroomLevel(15, t), 'yellow');
    assert.equal(headroomLevel(14, t), 'red');
    assert.equal(headroomLevel(-5, t), 'red');
  });
});

describe('oklchToRgb', () => {
  it('maps achromatic lightness to gray', () => {
    assert.deepEqual(oklchToRgb(1, 0, 0), [255, 255, 255]);
    assert.deepEqual(oklchToRgb(0, 0, 0), [0, 0, 0]);
  });

  it('returns null outside the sRGB gamut', () => {
    assert.equal(oklchToRgb(0.8, 0.4, 265), null);
  });

  it('keeps channels as integers in 0-255 (regression: int rounding before gamma)', () => {
    const rgb = oklchToRgb(0.8, 0.1, 255);
    assert.ok(
      rgb.every((c) => Number.isInteger(c) && c > 0 && c < 255),
      String(rgb),
    );
  });
});

describe('project colors', () => {
  it('every slot is readable on a dark terminal (WCAG AA 4.5:1 and then some)', () => {
    for (let slot = 0; slot < PROJECT_SLOTS; slot++) {
      const ratio = contrastRatio(slotColor(slot), TERMINAL_BG);
      assert.ok(ratio >= 9, `slot ${slot} contrast ${ratio.toFixed(1)}`);
    }
  });

  it('every slot is a distinct color', () => {
    const colors = new Set(Array.from({ length: PROJECT_SLOTS }, (_, i) => slotColor(i).join(',')));
    assert.equal(colors.size, PROJECT_SLOTS);
  });

  it('slots have similar perceived brightness', () => {
    const lums = Array.from({ length: PROJECT_SLOTS }, (_, i) => relativeLuminance(slotColor(i)));
    assert.ok(Math.max(...lums) / Math.min(...lums) < 1.4);
  });

  it('hashes names stably and case-insensitively', () => {
    assert.equal(projectSlot('techcentral'), projectSlot('techcentral'));
    assert.equal(projectSlot('TechCentral'), projectSlot('techcentral'));
    const slot = projectSlot('anything');
    assert.ok(Number.isInteger(slot) && slot >= 0 && slot < PROJECT_SLOTS);
  });

  it('honors overrides case-insensitively and wraps out-of-range values', () => {
    assert.equal(projectSlot('Website', { website: 4 }), 4);
    assert.equal(projectSlot('api', { api: 13 }), 1);
    assert.equal(projectSlot('api', { api: -1 }), 11);
  });
});
